src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -154,202 +110,179 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
+                    try
                     {
-                        if (!Enum.TryParse<TransportType>(transport.Transport, out var transportType))
-                        {
-                            Log.TransportNotSupported(_logger, transport.Transport);
-                            continue;
-                        }
+                        await _transport.StopAsync();
 
-                        if (transportType == TransportType.WebSockets && !IsWebSocketsSupported())
-                        {
-                            Log.WebSocketsNotSupportedByOperatingSystem(_logger);
-                            continue;
-                        }
-
-                        try
-                        {
-                            if ((transportType & _requestedTransportType) == 0)
-                            {
-                                Log.TransportDisabledByClient(_logger, transportType);
-                            }
-                            else if (!transport.TransferFormats.Contains(transferFormatString, StringComparer.Ordinal))
-                            {
-                                Log.TransportDoesNotSupportTransferFormat(_logger, transportType, transferFormat);
-                            }
-                            else
-                            {
-                                // The negotiation response gets cleared in the fallback scenario.
-                                if (negotiationResponse == null)
-                                {
-                                    negotiationResponse = await GetNegotiationResponse();
-                                    connectUrl = CreateConnectUrl(Url, negotiationResponse.ConnectionId);
-                                }
-
-                                Log.StartingTransport(_logger, transportType, connectUrl);
-                                await StartTransport(connectUrl, transportType, transferFormat);
-                                break;
-                            }
-                        }
-                        catch (Exception ex)
-                        {
-                            Log.TransportFailed(_logger, transportType, ex);
-                            // Try the next transport
-                            // Clear the negotiation response so we know to re-negotiate.
-                            negotiationResponse = null;
-                        }
+                        // Wait for the transport to complete its side of the pipes
+                        // REVIEW: We don't have to do this. But I think I do prefer that DisposeAsync only completes when everything is indeed complete
+                        // We could just unref the pipe and assume it'll shut down. All our transports will be tested to ensure they properly shut down though.
+                        await _transportPipe.Input.WaitForWriterToComplete();
+                        await _transportPipe.Output.WaitForReaderToComplete();
```

Comment:
Open Question: Should we wait for the transport to shut down in `DisposeAsync`? I think so. Related question: Should HubConnection wait for the HttpConnection to dispose? Right now I have it doing so, but we could just unref it and then fire-and-forget Dispose. The difference is in how bugs with HttpConnection shutdown manifest: If we wait for DisposeAsync, they become hangs, if we don't, memory leaks :)

Reply:
The HttpConnection should wait on the transport to die. The HubConnection should wait for a fixed amount of time for graceful close then cancel its own read loop after.

Reply:
Yep, that's what it does currently: https://github.com/aspnet/SignalR/blob/7cddcb5f7b829668588d9967a849722228773dab/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs#L780

Reply:
Ok I took another look at this and had flashbacks to when I tried.

- What should the behavior be if I create an HttpConnection, call StartAsync and then call DisposeAsync()? Currently it won't hang because you're forcefully completing the transport pipe no matter what. We need to make that part of the contract. DisposeAsync will complete the pipes so you cannot call it and then use the pipes.